### ISSUE TITLE
Add descriptions and examples to each lint rule

### ISSUE THREAD

**Issue by bartlomieju** [denoland/deno_lint#159]
All lint rules should be updated to contains some kind of method that returns description as well as some examples of valid and invalid. This feature will be used to generate documentation.

- [x] adjacent-overload-signatures
- [x] ban-ts-comment
- [x] ban-types
- [x] ban-untagged-ignore
- [x] ban-untagged-todo
- [x] constructor-super
- [x] default-param-last
- [x] eqeqeq
- [x] explicit-function-return-type
- [x] explicit-module-boundary-types
- [x] for-direction
- [x] getter-return
- [x] no-array-constructor
- [x] no-async-promise-executor
- [x] no-await-in-loop
- [x] no-case-declarations
- [x] no-class-assign
- [x] no-compare-neg-zero
- [x] no-cond-assign
- [x] no-const-assign
- [x] no-constant-condition
- [x] no-control-regex
- [x] no-debugger
- [x] no-delete-var
- [x] no-dupe-args
- [x] no-dupe-class-members
- [x] no-dupe-else-if
- [x] no-dupe-keys
- [x] no-duplicate-case
- [x] no-empty
- [x] no-empty-character-class
- [x] no-empty-interface
- [x] no-empty-pattern
- [x] no-eval
- [x] no-ex-assign
- [x] no-explicit-any
- [x] no-extra-boolean-cast
- [x] no-extra-non-null-assertion
- [x] no-extra-semi
- [x] no-fallthrough
- [x] no-func-assign
- [x] no-global-assign
- [x] no-import-assign
- [x] no-inferrable-types
- [x] no-inner-declarations
- [x] no-invalid-regexp
- [x] no-irregular-whitespace
- [x] no-misused-new
- [x] no-mixed-spaces-and-tabs
- [x] no-namespace
- [x] no-new-symbol
- [x] no-non-null-asserted-optional-chain
- [x] no-non-null-assertion
- [x] no-obj-calls
- [x] no-octal
- [x] no-prototype-builtins
- [x] no-redeclare
- [x] no-regex-spaces
- [x] no-self-assign
- [x] no-setter-return
- [x] no-shadow-restricted-names
- [x] no-sparse-arrays
- [x] no-this-alias
- [x] no-this-before-super
- [x] no-throw-literal
- [x] no-undef
- [x] no-unreachable
- [x] no-unsafe-finally
- [x] no-unsafe-negation
- [x] no-unused-labels
- [x] no-unused-vars
- [x] no-var
- [x] no-with
- [x] prefer-as-const
- [x] prefer-const
- [x] prefer-namespace-keyword
- [x] require-yield
- [x] single-var-declarator
- [x] triple-slash-reference
- [x] use-isnan
- [x] valid-typeof

**Comment by bartlomieju** [comment]
I think it's high time to start adding rules documentation so I'm pinning this issue.

So to support basic documentation and website generation we need to change `LintRule` trait minimally:
```rust
pub trait LintRule {
  fn new() -> Box<Self>
  where
    Self: Sized;
  fn lint_module(
    &self,
    context: Arc<Context>,
    module: &swc_ecmascript::ast::Module,
  );
  fn code(&self) -> &'static str;
  fn docs(&self) -> &'static str; // <--- new method
}
```

`LintRule::docs()` would return a string with written documentation. I think it makes sense to support markdown being returned from this function because we could print it out nice to terminal (using [`termimad`](https://github.com/Canop/termimad) for example) as well as auto-generate website similar to `doc.deno.land`.

Documentation for each lint rule should contain a short synopsis on what the rule does, why is it used and examples of both valid and invalid code.

CC @lucacasonato

**Comment by DomParfitt** [comment]
I'm happy to look at implementing this on the rules I've implemented once the trait is updated. From a quick scan that covers:
- `no-empty`
- `no-cond-assign`
- `valid-typeof`
- `no-unsafe-finally`
- `no-compare-neg-zero`
- `no-dupe-keys`

:slightly_smiling_face:

**Comment by bartlomieju** [comment]
@DomParfitt great, thank you! I've already merged https://github.com/denoland/deno_lint/pull/343
